src/safeds_stubgen/api_analyzer/_ast_visitor.py

```diff
@@ -329,68 +364,161 @@ def leave_assignmentstmt(self, _: AssignmentStmt) -> None:
+                        # Otherwise, we can parse the type normally
+                        unanalyzed_ret_type = getattr(node.unanalyzed_type, "ret_type", None)
+                        ret_type = mypy_type_to_abstract_type(node_ret_type, unanalyzed_ret_type)
+            else:
+                # Infer type
+                ret_type = self._infer_type_from_return_stmts(node)
+                is_type_inferred = ret_type is not None
 
         if ret_type is None:
             return []
 
-        results = []
-
-        docstring = self.docstring_parser.get_result_documentation(node)
-        if isinstance(ret_type, sds_types.TupleType):
-            for i, type_ in enumerate(ret_type.types):
-                name = f"result_{i + 1}"
-                results.append(
-                    Result(
-                        id=f"{function_id}/{name}",
-                        type=type_,
-                        name=name,
-                        docstring=docstring,
-                    ),
-                )
-        else:
-            name = "result_1"
-            results.append(
+        result_docstring = self.docstring_parser.get_result_documentation(node)
+
+        if is_type_inferred and isinstance(ret_type, sds_types.TupleType):
+            return self._create_infered_results(ret_type, result_docstring, function_id)
+
+        # If we got a TupleType, we can iterate it for the results, but if we got a NamedType, we have just one result
+        return_results = ret_type.types if isinstance(ret_type, sds_types.TupleType) else [ret_type]
+        return [
+            Result(
+                id=f"{function_id}/result_{i + 1}",
+                type=type_,
+                is_type_inferred=is_type_inferred,
+                name=f"result_{i + 1}",
+                docstring=result_docstring,
+            )
+            for i, type_ in enumerate(return_results)
+        ]
+
+    @staticmethod
+    def _infer_type_from_return_stmts(func_node: mp_nodes.FuncDef) -> sds_types.NamedType | sds_types.TupleType | None:
+        # To infer the type, we iterate through all return statements we find in the function
+        func_defn = get_funcdef_definitions(func_node)
+        return_stmts = find_return_stmts_recursive(func_defn)
+        if return_stmts:
+            types = {
+                mypy_expression_to_sds_type(return_stmt.expr)
+                for return_stmt in return_stmts
+                if return_stmt.expr is not None
+            }
+
+            # We have to sort the list for the snapshot tests
+            return_stmt_types = list(types)
+            return_stmt_types.sort(
+                key=lambda x: (x.name if isinstance(x, sds_types.NamedType) else str(len(x.types))),
+            )
+
+            if len(return_stmt_types) >= 2:
+                return sds_types.TupleType(types=return_stmt_types)
+            return return_stmt_types[0]
+        return None
+
+    @staticmethod
+    def _create_infered_results(
```

Comment:
```suggestion
    def _create_inferred_results(
```

Don't apply the suggestion directly but refactor.

Reply:
Feel free to resolve conversations where you are certain that they are handled.